sklearn/preprocessing/discretization.py

```diff
@@ -0,0 +1,217 @@
+           [ 1., 2., 1., 0.],
+           [ 2., 3., 2., 1.],
+           [ 2., 4., 2., 2.]])
+    """
+
+    def __init__(self, n_bins=2, categorical_features=None):
+        self.n_bins = n_bins
+        self.categorical_features = categorical_features
+
+        # Attributes
+        self.mins_ = None
+        self.maxes_ = None
+        self.n_features_ = None
+        self.cut_points_ = None
+        self.categorical_features_set_ = None
+
+    def fit(self, X, y=None):
+        """Fits the estimator.
+
+        Parameters
+        ----------
+        X : numeric array-like, shape (n_samples, n_features)
+            The input data containing features to be discretized. Indexes
+            of categorical features are indicated by categorical_features
+            from the constructor.
+
+        Returns
+        -------
+        self
+        """
+
+        if self.n_bins < 2:
+            raise ValueError("Discretizer received an invalid number "
+                             "of bins. Received {0}, expected at least 2."
+                             .format(self.n_bins))
+
+        X = check_array(X, dtype=float, ensure_2d=False)
+
+        if X.ndim == 1:
+            raise ValueError("Input array must be 2D shape.")
+
+        self._check_categorical_features(self.categorical_features, X.shape[1])
+
+        self._fit(X)
+        return self
+
+    @staticmethod
+    def _check_categorical_features(cat_features, n_features):
+        """Only allow integer indices, or None.
+        """
+        if cat_features is None:
+            return
+
+        cat_features = check_array(cat_features, ensure_2d=False, dtype=int)
+        cat_features = column_or_1d(cat_features)
+
+        if len(set(cat_features)) != cat_features.shape[0]:
+            raise ValueError("Duplicate categorical column indexes found.")
+
+        if np.all(cat_features >= 0) and np.all(cat_features < n_features):
+            return
+
+        raise ValueError("Invalid categorical feature index.")
+
+    def _cut_points_from_index(self, col_idx):
+        """Determines the cut points of a feature based on the column index.
+        """
+        n_bins = self.n_bins
+        if col_idx in self.categorical_features_set_:
+            return [np.nan] * (self.n_bins - 1)
+
+        _min = self.mins_[col_idx]
+        _max = self.maxes_[col_idx]
+        spacing = (_max - _min) / n_bins
+        return np.linspace(_min + spacing, _max - spacing, num=n_bins - 1)
+
+    def _fit(self, X):
+        """Fits the attributes of the discretizer.
+        """
+        n_features = X.shape[1]
+        if self.categorical_features:
+            self.categorical_features_set_ = set(self.categorical_features)
+        else:
+            self.categorical_features_set_ = {}
+
+        self.mins_ = np.min(X, axis=0)
+        self.maxes_ = np.max(X, axis=0)
+
+        if np.any(self.mins_ - self.maxes_ == 0):
+            raise ValueError("Array contained features which contained the "
```

Comment:
I think this should be at best a warning. everything goes into one bin in this case.